Now library actually compiled in no-std

LucaCappelletti94 opened the pull request:
I noticed that while the library has an "std" feature that can be turned off, it does not compile without std and alloc.

Now it compiles and provides a reduced set of features (e.g. MemSize).

I have also added tests in the CI to verify that it builds and runs the tests successfully also in no-std mode.

vigna commented:
Sorry for the delay.

I think you left the local path in the mem_dbg-derive dependency.

Is the reduction in features documented in the README? Which features are missing? Just support for structures such as `HashMap`, or more?

vigna commented:
Any chance you're gonna update this?

vigna commented:
Very nice! However, as in other crates, we prefer to have an `[[example]]` section in `Cargo.tml` with a `required-features` item rather than conditional compilation of the main method. Less code, and the user knows at compilation time. E.g., from `sux`:

```
[[example]]
name = "bench_vfunc"
required-features = ["epserde"]
```

vigna commented:
Yes, but the gating will appear in the home page of `crates.io`. Not a big issue.

vigna commented:
We do it here: https://crates.io/crates/dsi-bitstream

LucaCappelletti94 commented:
Ok so, I extended the test coverage and uncovered a pandora's box worth of bugs. I fixed some, but several are in smart pointers and other types that I am not too familiar with, so I will now push the edits with the new (failing) tests. Also, I have:

1. Removed the broken impl of Arc
2. Resolved some subtraction underflow errors, but I am not certain whether those memsize impls were correct in the first place
3. Fixed some plainly incorrect memsize impls

vigna commented:
Feel free to comment the subsequent modifications...
